File: dynachart/dynachart.pl
#!/usr/bin/env perl

# dynachart.pl
# Jared Still 2017-07-23
# still@pythian.com jkstill@gmail.com

# data is from STDIN

use warnings;
use strict;
use Data::Dumper;
use Pod::Usage;
use Getopt::Long;
use Excel::Writer::XLSX;

my $debug = 0;
my $combinedChart = 0;
my %optctl = ();
my ($help,$man);
my @chartCols;
my $categoryColName='';
my $categoryColNum=0;
my $secondaryAxisCol='';
my $autoFilterEnabled=1;

my $chartType='line';
my $chartTitles=1;
my @chartTypesAvailable=qw(area bar column line pie doughnut scatter stock);
my @legendPositions=qw(bottom left right top);
my $legendPosition='bottom';

my $delimiter=','; # default delimiter of comma
my $listAvailableColumns=0;

Getopt::Long::GetOptions(
	\%optctl, 
	'spreadsheet-file=s',
	'debug!' => \$debug,
	'chart-cols=s{1,10}' => \@chartCols,
	'chart-type=s' => \$chartType,
	'secondary-axis-col=s' => \$secondaryAxisCol,
	'auto-filter-enabled!' => \$autoFilterEnabled,
	'legend-position=s' => \$legendPosition,
	'combined-chart!' => \$combinedChart,
	'chart-titles!' => \$chartTitles,
	'list-available-columns!' => \$listAvailableColumns,
	'worksheet-col=s',  # creates a separate worksheet per value of this column
	'category-col=s' => \$categoryColName,
	'delimiter=s' => \$delimiter,
	'h|help|?' => \$help, man => \$man
) or pod2usage(2) ;

pod2usage(1) if $help;
pod2usage(-verbose => 2) if $man;


unless ( grep { $_ eq $chartType } @chartTypesAvailable) { 
	print qq {

$chartType is not valid chart type

};
	pod2usage(1);
}


my $xlFile = defined($optctl{'spreadsheet-file'}) ? $optctl{'spreadsheet-file'} : 'asm-metrics.xlsx';
my $workSheetCol = defined($optctl{'worksheet-col'}) ? $optctl{'worksheet-col'} : 0;

my %fonts = (
	fixed			=> 'Courier New',
	fixed_bold	=> 'Courier New',
	text			=> 'Arial',
	text_bold	=> 'Arial',
);

my %fontSizes = (
	fixed			=> 10,
	fixed_bold	=> 10,
	text			=> 10,
	text_bold	=> 10,
);

my $maxColWidth = 50;
my $counter = 0;
my $interval = 100;

# create workbook
my $workBook = Excel::Writer::XLSX->new($xlFile);
die "Problems creating new Excel file $xlFile: $!\n" unless defined $workBook;

# create formats
my $stdFormat = $workBook->add_format(bold => 0,font => $fonts{fixed}, size => $fontSizes{fixed}, color => 'black');
my $boldFormat = $workBook->add_format(bold => 1,font => $fonts{fixed_bold}, size => $fontSizes{fixed_bold}, color => 'black');
my $wrapFormat = $workBook->add_format(bold => 0,font => $fonts{text}, size => $fontSizes{text}, color => 'black');
$wrapFormat->set_align('vjustify');


my $labels=<>;
#chomp $labels;
# windows created files may have CR
# chomp on linux will not remove the CR
$labels =~ s/[\r\n]+$//;
# sadf starts header lines with '# ' - remove that
$labels =~ s/^#\s+//;
my @labels = split(/$delimiter\s*/,$labels);

if ($listAvailableColumns) {

	print join("\n",@labels),"\n";
	exit;

}


if ($debug) {

print qq{LABELS:\n};

print join("\n",@labels);

print "\n";

}

# get the X series category
if ( $categoryColName ) {

	my $want = $categoryColName;
	my $index = 0;
	++$index until ($labels[$index] eq $want) or ($index > $#labels);
	$categoryColNum = $index;	

}

#print join("\n",@labels);

# get the element number of the column used to segregate into worksheets
my $workSheetColPos;
if ($workSheetCol) {
	my $i=0;
	foreach my $label ( @labels)  {
		if ($label eq $workSheetCol) { 
			$workSheetColPos = $i;
			last;
		}
		$i++;
	}
}

print "\nworkSheetColPos: $workSheetColPos\n" if $debug;

# validate the columns to be charted
# use as an index into the labels
my @chartColPos=();
{
	my $i=0;
	print "validating chart columns\n" if $debug;
	foreach my $label ( @labels)  {
		foreach my $chartCol ( @chartCols ) {
			print "label: |$label|\nchartCol: |$chartCol|\n============\n" if $debug;
			if ($label eq $chartCol) { 
				print "Pushing $i to \@chartColPos\n" if $debug;
				push @chartColPos, $i;
				last;
			}
		}
		$i++;
	}
}

# validate secondary axis column
print "validating secondary axis column\n" if $debug;
if ($secondaryAxisCol ) {
	my $foundSecondaryAxisCol=0;
	foreach my $label ( @labels) {
		if ($label eq $secondaryAxisCol) {
			$foundSecondaryAxisCol = 1;
			last;
		}
	}

	if ( ! $foundSecondaryAxisCol ) {
		warn "\n secondary axis column of '$secondaryAxisCol' is invalid - not using secondary axis\n";
		$secondaryAxisCol='';
	}
}

if ($debug) {
	print "\nworkSheetCol:\n", Dumper(\$workSheetCol);
	print "\nChartCols:\n", Dumper(\@chartCols);
	print "\nChartColPos:\n", Dumper(\@chartColPos);
	print "\nLabels:\n", Dumper(\@labels);
}

my %lineCount=();
my %workSheets=();


# the first worksheet is a directory
my $directoryName='Directory';
my $directory;
my $noDirectory=0;
my $directoryLineCount=0;

unless ($noDirectory) {
	$directory = $workBook->add_worksheet($directoryName)	;
	$directory->set_column(0,0,30);
	$directory->write_row($directoryLineCount++,0,['Directory'],$boldFormat);
}

while (<STDIN>) {

	#chomp; 
	# windows created files may have CR
	# chomp on linux will not remove the CR
	s/[\r\n]+$//;
	my @data=split(/$delimiter/);

	my $currWorkSheetName;
	if ($workSheetCol) {
		$currWorkSheetName=$data[$workSheetColPos];
		if (length($currWorkSheetName) > 31 ) {
			# cut some out of the middle of the name as the Excel worksheet name has max size of 31
			$currWorkSheetName = substr($currWorkSheetName,0,14) . '...' . substr($currWorkSheetName,-14);
		}
	} else {
		$currWorkSheetName='DynaChart';
	}

	print "Worksheet Name: $currWorkSheetName\n" if $debug;

	unless (defined $workSheets{$currWorkSheetName}) {
		$workSheets{$currWorkSheetName} = $workBook->add_worksheet($currWorkSheetName);
		$workSheets{$currWorkSheetName}->write_row($lineCount{$currWorkSheetName}++,0,\@labels,$boldFormat);
		# freeze pane at header
		$workSheets{$currWorkSheetName}->freeze_panes($lineCount{$currWorkSheetName},0);
		# autofilter
		$workSheets{$currWorkSheetName}->autofilter(0,0,0,$#data) if $autoFilterEnabled;
	}

	# setup column widths
	#$workSheet->set_column($el,$el,$colWidth);
	$workSheets{$currWorkSheetName}->write_row($lineCount{$currWorkSheetName}++,0,\@data, $stdFormat);

}

if ($debug) {
	print "Worksheets:\n";
	print "$_\n" foreach keys %workSheets;
	print Dumper(\%lineCount);
}

# each row consumes about 18 pixels
my $rowHeight=18; # pixels
my $chartHeight=23; # rows

# triple from default width of 480
my $chartWidth = 480 * 3;

=head1 Write the Charts

 The default mode is to create a separate chart for each metric
 By specifying the command line option --combined-chart the values will be charted in a single chart
 Doing so is probably useful only for a limited number of sets of values

 Some may question the apparent duplication of code in the sections to combine or not combine the charts
 Doing so would be a bit convoluted - this is easier to read and modify

=cut

foreach my $workSheet ( keys %workSheets ) {
	print "Charting worksheet: $workSheet\n" if $debug;

	my $chartNum = 0;

	if ($combinedChart) {
		my $chart = $workBook->add_chart( type => $chartType, name => "Combined" . '-' . $workSheets{$workSheet}->get_name(), embedded => 1 );
		$chart->set_size( width => $chartWidth, height => $chartHeight * $rowHeight);
		$chart->set_legend( position => $legendPosition );

		# combined chart should only be 2 columns
		if ($chartTitles) {
			my $chartTitleCols = '';
			foreach my $colPos ( @chartColPos ) {
				$chartTitleCols .= $labels[$colPos] . '/';
			}
			chop $chartTitleCols;

			$chart->set_title( name => $workSheets{$workSheet}->get_name() . ' - ' . $chartTitleCols );
		}
		
		# each chart consumes about 16 rows
		$workSheets{$workSheet}->insert_chart((($chartNum * $chartHeight) + 2),3, $chart);

		foreach my $colPos ( @chartColPos ) {
			my $col2Chart=$labels[$colPos];
			# [ sheet, row_start, row_end, col_start, col_end]
			$chart->add_series(
				name => $col2Chart,
				#categories => [$workSheet, 1,$lineCount{$workSheet},2,2],
				categories => [$workSheet, 1,$lineCount{$workSheet},$categoryColNum,$categoryColNum],
				values => [$workSheet, 1,$lineCount{$workSheet},$colPos,$colPos],
				y2_axis => $col2Chart eq $secondaryAxisCol ? 1 : 0
			);
		}
		
	} else {
		foreach my $colPos ( @chartColPos ) {
			my $col2Chart=$labels[$colPos];
			print "\tCharting column: $col2Chart\n" if $debug;
			my $chart = $workBook->add_chart( type => $chartType, name => "$col2Chart" . '-' . $workSheets{$workSheet}->get_name(), embedded => 1 );
			$chart->set_size( width => $chartWidth, height => $chartHeight * $rowHeight);
			$chart->set_legend( position => $legendPosition );
			$chart->set_title( name => $workSheets{$workSheet}->get_name() . ' - ' . $labels[$colPos]  ) if $chartTitles;

			# each chart consumes about 16 rows
			$workSheets{$workSheet}->insert_chart((($chartNum * $chartHeight) + 2),3, $chart);
		

			# [ sheet, row_start, row_end, col_start, col_end]
			$chart->add_series(
				name => $col2Chart,
				#categories => [$workSheet, 1,$lineCount{$workSheet},2,2],
				categories => [$workSheet, 1,$lineCount{$workSheet},$categoryColNum,$categoryColNum],
				values => [$workSheet, 1,$lineCount{$workSheet},$colPos,$colPos]
			);

			$chartNum++;
		}
	}
}


# write the directory page
my $urlFormat = $workBook->add_format( color => 'blue', underline => 1 );
my %sheetNames=();

foreach my $worksheet ( $workBook->sheets() ) {
	my $sheetName = $worksheet->get_name();
	next if $sheetName eq $directoryName;
	$sheetNames{$sheetName} = $worksheet;
}

foreach my $sheetName ( sort keys %sheetNames ) {
	$directory->write_url($directoryLineCount++,0, qq{internal:'$sheetName'!A1} ,$urlFormat, $sheetName);
}

__END__

=head1 NAME

dynachart.pl

  --help brief help message
  --man  full documentation
  --spreadsheet-file output file name - defaults to asm-metrics.xlsx
  --worksheet-col name of column used to segragate data into worksheets 
    defaults to a single worksheet if not supplied
  --legend-position     left, right, top, bottom - default is bottom
  --auto-filter-enabled enable the drop down Excel filters
  --chart-type default chart type is 'line'
  --chart-cols list of columns to chart
  --chart-titles       enable chart titles (default)
  --secondary-axis-col name of the column to be on secondary axis
                       works only with combined-chart option
  --legend-position     left, right, top, bottom - default is bottom
  --chart-titles       enable chart titles (default)
  --secondary-axis-col name of the column to be on secondary axis
                       works only with combined-chart option
  --legend-position     left, right, top, bottom - default is bottom
  --auto-filter-enabled enable the drop down Excel filters

 dynachart.pl accepts input from STDIN

 This script will read CSV data created by asm-metrics-collector.pl or asm-metrics-aggregator.pl


=head1 SYNOPSIS

dynachart.pl [options] [file ...]

 Options:
   --help                  brief help message
   --man                   full documentation
   --spreadsheet-file      output file name - defaults to asm-metrics.xlsx
   --worksheet-col         name of column used to segragate data into worksheets 
                           defaults to a single worksheet if not supplied
  --chart-type             default chart type is 'line'
  --chart-cols             list of columns to chart
  --chart-titles           enable chart titles (default)
                           worksheet name and column name
							      combined chart: worksheet name and column names
  --secondary-axis-col     name of the column to be on secondary axis
                           works only with combined-chart option
  --legend-position        left, right, top, bottom - default is bottom
  --auto-filter-enabled    enable the drop down Excel filters
  --category-col specify   the column for the X vector - a timestamp is typically used 
                           the name must exactly  match that in the header
  --combined-chart         create a single chart rather than a chart for each value specified in --chart-cols

 dynachart.pl accepts input from STDIN

 dynachart.pl --worksheet-col DISKGROUP_NAME < my_input_file.csv


 dynachart.pl --spreadsheet-file sar-disk-test.xlsx --combined-chart --worksheet-col DEV --category-col 'timestamp' --chart-cols 'rd_sec/s' --chart-cols 'wr_sec/s' < sar-disk-test.csv


=head1 OPTIONS

=over 8

=item B<-help>

Print a brief help message and exits.

=item B<-man>

Prints the manual page and exits.

=item B<--spreadsheet-file>

 The name of the Excel file to create.
 The default name is asm-metrics.xlsx

=item B<--worksheet-col>

 By default a single worksheet is created.
 When this option is used the column supplied as an argument will be used to segragate data into separate worksheets.

=item B<--chart-type>

 The valid charts are 
  area 
  bar 
  column 
  line 
  pie 
  doughnut 
  scatter 
  stock

 The default chart type is 'line'

=item B<--chart-cols>

 List of columns to chart
 This should be the last option on the command line if used.

 It may be necessary to tell Getopt to stop processing arguments with '--' in some cases.

 eg.

 dynachart.pl dynachart.pl --worksheet-col DISKGROUP_NAME --chart-cols READS WRITES -- logs/asm-oravm-20150512_01-agg-dg.csv

=item B<--category-col>

 Column to use as the category for the X line in the chart - default to the first column
 The name must exactly match a column from the CSV file
 Typically this line is a timestamp

=item B<--secondary-axis-col>

 Use a secondary axis for one of the series columns

   --chart-cols TIME --chart-cols READS --chart-col BYTES_READ \
   --secondary-axis-col BYTES_READ 

=item B<--auto-filter-enabled>

  Enable the drop down Excel filters

  Default is ON.  
  Disable with --noauto-filter-enabled

=item B<--delimiter>

 The default input delimiter is a comma - ,
 Change the delimiter to some other value

 eg. change to a colon
    --delimiter :

=item B<--list-available-columns>

 List the columns from the first line of the file and exit

=back

=head1 DESCRIPTION

B<dynachart.pl> creates an excel file with charts for selected columns>

=head1 EXAMPLES

 dynachart.pl accepts data from STDIN
 
 dynachart.pl --worksheet-col DISKGROUP_NAME --spreadsheet-file mywork.xlsx

=cut

File: dynachart/dynachart.py
#!/usr/bin/env python3

# dynachart.py
# Jared Still 2017-07-23 (original Perl version)
# Python conversion 2026-04-13
# still@pythian.com jkstill@gmail.com

# Data is read from STDIN

import sys
import re
import argparse
import xlsxwriter


CHART_TYPES_AVAILABLE = ['area', 'bar', 'column', 'line', 'pie', 'doughnut', 'scatter', 'stock']
LEGEND_POSITIONS = ['bottom', 'left', 'right', 'top']


def parse_args():
    parser = argparse.ArgumentParser(
        prog='dynachart.py',
        description='Create an Excel file with charts from CSV data read on STDIN.',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
CHART TYPES:
  area, bar, column, line (default), pie, doughnut, scatter, stock

LEGEND POSITIONS:
  bottom (default), left, right, top

EXAMPLES:

  dynachart.py accepts data from STDIN

  dynachart.py --worksheet-col DISKGROUP_NAME --spreadsheet-file mywork.xlsx

  dynachart.py --spreadsheet-file sar-disk-test.xlsx --combined-chart \\
    --worksheet-col DEV --category-col timestamp \\
    --chart-cols rd_sec/s wr_sec/s < sar-disk-test.csv
        """,
    )

    parser.add_argument(
        '--spreadsheet-file',
        default='asm-metrics.xlsx',
        metavar='FILE',
        help='Output Excel file name (default: asm-metrics.xlsx)',
    )
    parser.add_argument(
        '--debug',
        action=argparse.BooleanOptionalAction,
        default=False,
        help='Enable debug output (default: off)',
    )
    parser.add_argument(
        '--chart-cols',
        nargs='+',
        action='append',
        default=None,
        metavar='COL',
        help='One or more column names to chart; may be repeated',
    )
    parser.add_argument(
        '--chart-type',
        default='line',
        choices=CHART_TYPES_AVAILABLE,
        help='Chart type (default: line)',
    )
    parser.add_argument(
        '--secondary-axis-col',
        default='',
        metavar='COL',
        help='Column name to plot on secondary Y axis (combined chart only)',
    )
    parser.add_argument(
        '--auto-filter-enabled',
        action=argparse.BooleanOptionalAction,
        default=True,
        help='Enable Excel autofilter dropdowns (default: on; disable with --no-auto-filter-enabled)',
    )
    parser.add_argument(
        '--legend-position',
        default='bottom',
        choices=LEGEND_POSITIONS,
        help='Chart legend position (default: bottom)',
    )
    parser.add_argument(
        '--combined-chart',
        action=argparse.BooleanOptionalAction,
        default=False,
        help='Create one combined chart for all --chart-cols instead of one chart per column',
    )
    parser.add_argument(
        '--chart-titles',
        action=argparse.BooleanOptionalAction,
        default=True,
        help='Add titles to charts (default: on; disable with --no-chart-titles)',
    )
    parser.add_argument(
        '--list-available-columns',
        action=argparse.BooleanOptionalAction,
        default=False,
        help='Print column names from the header line and exit',
    )
    parser.add_argument(
        '--worksheet-col',
        default='',
        metavar='COL',
        help='Column name whose values are used to segregate data into separate worksheets',
    )
    parser.add_argument(
        '--category-col',
        default='',
        metavar='COL',
        help='Column name to use as the X-axis category (typically a timestamp); defaults to first column',
    )
    parser.add_argument(
        '--delimiter',
        default=',',
        help='Input field delimiter (default: comma)',
    )

    return parser.parse_args()


def truncate_sheet_name(name):
    """Excel worksheet names are limited to 31 characters.
    When the name is longer, cut out the middle to preserve both ends."""
    if len(name) > 31:
        return name[:14] + '...' + name[-14:]
    return name


def coerce_numeric(val):
    """Convert a string to int or float if possible, otherwise return it as-is.
    xlsxwriter writes whatever Python type it receives, so without this
    conversion every split() value would be stored as text in Excel."""
    try:
        return int(val)
    except ValueError:
        try:
            return float(val)
        except ValueError:
            return val


def main():
    args = parse_args()

    debug               = args.debug
    combined_chart      = args.combined_chart
    # action='append' + nargs='+' produces a list-of-lists; flatten it.
    # Supports both  --chart-cols a b c  and  --chart-cols a --chart-cols b
    chart_cols = [c for sublist in (args.chart_cols or []) for c in sublist]
    category_col_name   = args.category_col
    secondary_axis_col  = args.secondary_axis_col
    auto_filter_enabled = args.auto_filter_enabled
    chart_type          = args.chart_type
    chart_titles        = args.chart_titles
    legend_position     = args.legend_position
    delimiter           = args.delimiter
    xl_file             = args.spreadsheet_file
    worksheet_col       = args.worksheet_col
    list_available_columns = args.list_available_columns

    # ------------------------------------------------------------------
    # Read and parse the header line
    # ------------------------------------------------------------------
    header_line = sys.stdin.readline()
    if not header_line:
        print("Error: no input data on STDIN", file=sys.stderr)
        sys.exit(1)

    # Strip CR/LF (handle Windows line endings)
    header_line = header_line.rstrip('\r\n')

    # sadf starts header lines with '# ' — remove that prefix
    header_line = re.sub(r'^#\s+', '', header_line)

    # Split on delimiter followed by optional whitespace (matches Perl behaviour)
    labels = re.split(re.escape(delimiter) + r'\s*', header_line)

    if list_available_columns:
        print('\n'.join(labels))
        sys.exit(0)

    if debug:
        print("LABELS:")
        print('\n'.join(labels))
        print()

    # ------------------------------------------------------------------
    # Resolve the category (X-axis) column index
    # ------------------------------------------------------------------
    category_col_num = 0
    if category_col_name:
        if category_col_name in labels:
            category_col_num = labels.index(category_col_name)
        else:
            print(
                f"Warning: category column '{category_col_name}' not found in headers",
                file=sys.stderr,
            )

    # ------------------------------------------------------------------
    # Resolve the worksheet-segregation column position
    # ------------------------------------------------------------------
    worksheet_col_pos = None
    if worksheet_col:
        if worksheet_col in labels:
            worksheet_col_pos = labels.index(worksheet_col)
        else:
            print(
                f"Warning: worksheet column '{worksheet_col}' not found in headers",
                file=sys.stderr,
            )
            worksheet_col = ''

    if debug:
        print(f"worksheet_col_pos: {worksheet_col_pos}")

    # ------------------------------------------------------------------
    # Resolve chart column positions (preserves the order given on CLI)
    # ------------------------------------------------------------------
    chart_col_pos = []
    missing_chart_cols = []
    for col in chart_cols:
        if col in labels:
            chart_col_pos.append(labels.index(col))
        else:
            missing_chart_cols.append(col)

    if missing_chart_cols:
        print(
            f"Warning: the following chart columns were not found in the data: "
            f"{', '.join(missing_chart_cols)}",
            file=sys.stderr,
        )

    if chart_cols and not chart_col_pos:
        print(
            "Warning: none of the requested chart columns are available — "
            "the spreadsheet will be created without charts.",
            file=sys.stderr,
        )

    # ------------------------------------------------------------------
    # Validate secondary-axis column
    # ------------------------------------------------------------------
    if secondary_axis_col and secondary_axis_col not in labels:
        print(
            f"\n secondary axis column of '{secondary_axis_col}' is invalid"
            " - not using secondary axis",
            file=sys.stderr,
        )
        secondary_axis_col = ''

    if debug:
        print(f"worksheet_col: {worksheet_col}")
        print(f"chart_cols: {chart_cols}")
        print(f"chart_col_pos: {chart_col_pos}")
        print(f"labels: {labels}")

    # ------------------------------------------------------------------
    # Create workbook and formats
    # ------------------------------------------------------------------
    workbook = xlsxwriter.Workbook(xl_file)

    std_format = workbook.add_format({
        'bold': False,
        'font_name': 'Courier New',
        'font_size': 10,
        'font_color': 'black',
    })
    bold_format = workbook.add_format({
        'bold': True,
        'font_name': 'Courier New',
        'font_size': 10,
        'font_color': 'black',
    })

    # ------------------------------------------------------------------
    # Directory worksheet (first sheet — an index with hyperlinks)
    # ------------------------------------------------------------------
    directory_name = 'Directory'
    directory = workbook.add_worksheet(directory_name)
    directory.set_column(0, 0, 30)
    directory_row = 0
    directory.write_row(directory_row, 0, ['Directory'], bold_format)
    directory_row += 1

    # ------------------------------------------------------------------
    # Read data rows and populate worksheets
    # ------------------------------------------------------------------
    # line_counts[name] tracks the next available row index for each sheet
    line_counts = {}
    worksheets  = {}

    for line in sys.stdin:
        line = line.rstrip('\r\n')
        data = line.split(delimiter)

        # Determine which worksheet this row belongs to
        if worksheet_col and worksheet_col_pos is not None:
            curr_name = truncate_sheet_name(data[worksheet_col_pos])
        else:
            curr_name = 'DynaChart'

        if debug:
            print(f"Worksheet Name: {curr_name}")

        # Create the worksheet on first encounter
        if curr_name not in worksheets:
            ws = workbook.add_worksheet(curr_name)
            worksheets[curr_name] = ws
            line_counts[curr_name] = 0

            ws.write_row(line_counts[curr_name], 0, labels, bold_format)
            line_counts[curr_name] += 1

            # Freeze pane below the header row
            ws.freeze_panes(line_counts[curr_name], 0)

            # Autofilter on the header row
            if auto_filter_enabled:
                ws.autofilter(0, 0, 0, len(data) - 1)

        worksheets[curr_name].write_row(
            line_counts[curr_name], 0,
            [coerce_numeric(v) for v in data],
            std_format,
        )
        line_counts[curr_name] += 1

    if debug:
        print("Worksheets:", list(worksheets.keys()))
        print("line_counts:", line_counts)

    # ------------------------------------------------------------------
    # Create charts (skipped when no valid chart columns were resolved)
    #
    # Default mode: one chart per column listed in --chart-cols.
    # With --combined-chart: a single chart containing all series.
    #
    # Chart dimensions mirror the original Perl values:
    #   width  = 480 * 3 = 1440 px
    #   height = 23 rows * 18 px/row = 414 px
    # ------------------------------------------------------------------
    row_height   = 18   # pixels per row
    chart_height = 23   # chart height in rows
    chart_width  = 480 * 3

    for ws_name, ws in (worksheets.items() if chart_col_pos else []):
        if debug:
            print(f"Charting worksheet: {ws_name}")

        # Data rows span rows 1 .. line_counts[ws_name]-1  (0-indexed)
        data_last_row = line_counts[ws_name] - 1

        chart_num = 0

        if combined_chart:
            chart = workbook.add_chart({'type': chart_type})
            chart.set_size({'width': chart_width, 'height': chart_height * row_height})
            chart.set_legend({'position': legend_position})

            if chart_titles:
                title_cols = '/'.join(labels[pos] for pos in chart_col_pos)
                chart.set_title({'name': f"{ws_name} - {title_cols}"})

            ws.insert_chart((chart_num * chart_height) + 2, 3, chart)

            for col_pos in chart_col_pos:
                col_name = labels[col_pos]
                # xlsxwriter list format: [sheet, first_row, first_col, last_row, last_col]
                series = {
                    'name':       col_name,
                    'categories': [ws_name, 1, category_col_num, data_last_row, category_col_num],
                    'values':     [ws_name, 1, col_pos,          data_last_row, col_pos],
                }
                if col_name == secondary_axis_col:
                    series['y2_axis'] = True
                chart.add_series(series)

        else:
            for col_pos in chart_col_pos:
                col_name = labels[col_pos]
                if debug:
                    print(f"\tCharting column: {col_name}")

                chart = workbook.add_chart({'type': chart_type})
                chart.set_size({'width': chart_width, 'height': chart_height * row_height})
                chart.set_legend({'position': legend_position})

                if chart_titles:
                    chart.set_title({'name': f"{ws_name} - {col_name}"})

                ws.insert_chart((chart_num * chart_height) + 2, 3, chart)

                # xlsxwriter list format: [sheet, first_row, first_col, last_row, last_col]
                chart.add_series({
                    'name':       col_name,
                    'categories': [ws_name, 1, category_col_num, data_last_row, category_col_num],
                    'values':     [ws_name, 1, col_pos,          data_last_row, col_pos],
                })

                chart_num += 1

    # ------------------------------------------------------------------
    # Write the Directory sheet — sorted hyperlinks to every data sheet
    # ------------------------------------------------------------------
    url_format = workbook.add_format({'font_color': 'blue', 'underline': 1})

    for sheet_name in sorted(worksheets.keys()):
        directory.write_url(
            directory_row, 0,
            f"internal:'{sheet_name}'!A1",
            url_format,
            sheet_name,
        )
        directory_row += 1

    workbook.close()


if __name__ == '__main__':
    main()
